download/download_generator.go

```diff
@@ -0,0 +1,201 @@
+}
+
+// GeneratorError is a wrapper for errors returned from the Generator
+type GeneratorError struct {
+	originalErr error
+	message     string
+	args        []interface{}
+}
+
+func newGeneratorError(err error, message string, args ...interface{}) GeneratorError {
+	return GeneratorError{
+		originalErr: err,
+		message:     message,
+		args:        args,
+	}
+}
+
+// Error return details about the error
+func (genErr GeneratorError) Error() string {
+	if genErr.originalErr == nil {
+		return errors.Errorf(genErr.message, genErr.args...).Error()
+	}
+	return errors.Wrap(genErr.originalErr, fmt.Sprintf(genErr.message, genErr.args...)).Error()
+}
+
+// GenerateDatasetDownloads generate unfiltererd download files for the specified dataset/edition/version
+func (g Generator) GenerateDatasetDownloads(datasetID string, edition string, versionID string, version string) error {
+	if err := g.validate(datasetID, edition, versionID, version); err != nil {
+		return err
+	}
+
+	versionURL := versionURI(datasetID, edition, version)
+	filterID, err := g.FilterClient.CreateBlueprint(versionID, []string{})
+	if err != nil {
+		return newGeneratorError(err, createBlueprintErr, versionURL)
+	}
+
+	jobState, err := g.FilterClient.GetJobState(filterID)
+	if err != nil {
+		return newGeneratorError(err, getJobStateErr, filterID, versionURL)
+	}
+
+	updatedBlueprint, err := g.FilterClient.UpdateBlueprint(jobState, true)
+	if err != nil {
+		return newGeneratorError(err, updateBlueprintErr, jobState.FilterID, versionURL)
+	}
+
+	filterOutput, err := g.checkDownloadsAvailable(updatedBlueprint.Links.FilterOutputs.ID, versionURL)
+	if err != nil {
+		return err
+	}
+
+	log.Info("full dataset version downloads available", log.Data{
+		downloadsKey: filterOutput.Downloads,
+		datasetIDKey: datasetID,
+		editionKey:   edition,
+		versionKey:   version,
+	})
+
+	latestVersion, err := g.Store.GetVersion(datasetID, edition, version, models.AssociatedState)
+	if err != nil {
+		return newGeneratorError(err, getVersionErr, versionURL)
+	}
+
+	xls := filterOutput.Downloads[xlsKey]
+	csv := filterOutput.Downloads[csvKey]
+
+	latestVersion.Downloads = &models.DownloadList{
+		XLS: &models.DownloadObject{URL: xls.URL, Size: xls.Size},
+		CSV: &models.DownloadObject{URL: csv.URL, Size: csv.Size},
+	}
+
+	if err := g.Store.UpdateVersion(latestVersion.ID, latestVersion); err != nil {
+		return newGeneratorError(err, updateDatasetVersionErr, versionURL)
+	}
+	return nil
+}
+
+func (g Generator) checkDownloadsAvailable(outputID string, versionURL string) (*filter.Model, error) {
+	var filterOutput filter.Model
+	var err error
+	available := false
+
+	for attempts := 0; attempts < g.MaxRetries; attempts++ {
+		filterOutput, err = g.FilterClient.GetOutput(outputID)
+		if err != nil {
+			log.Error(newGeneratorError(err, getOutputErr, outputID), nil)
+			continue
+		}
+		if g.downloadsAvailable(filterOutput) {
+			available = true
+			break
+		}
+		time.Sleep(g.RetryDelay)
```

Comment:
I think its better sleep at the start of the for loop. As the last sleep in the for loop won't do anything.